src/rstar.jl

```diff
@@ -161,7 +160,9 @@ function rstar(classif::MLJModelInterface.Supervised, x::AbstractArray{<:Any,3};
 end
 
 # R⋆ for deterministic predictions (algorithm 1)
-function _rstar(predictions::AbstractVector{T}, ytest::AbstractVector{T}) where {T}
+function _rstar(
+    ::MLJModelInterface.Deterministic, predictions::AbstractVector, ytest::AbstractVector
```

Comment:
Since we only support `Deterministic` and `Probabilistic`, perhaps we should constrain the types for `rstar` and update the docstring to only take `Union{MLJModelInterface.Probabilistic,MLJModelInterface.Deterministic}`.

Reply:
Perhaps not. And I'm not certain how common the other subtypes of `Supervised` are.

Reply:
We could also use StatisticalTraits, e.g., `StatisticalTraits.prediction_type`, throw a descriptive error if a `rstar` is called with a model for which it is not `:probabilistic` or `:deterministic` (see https://github.com/JuliaAI/MLJModelInterface.jl/blob/d9e9703947fc04b0a5e63680289e41d0ba0d65bd/src/model_traits.jl#L27-L28), and dispatch on it (using `Val`, it seems all these traits return `Symbol`s but since they are based on the type of the models, the compiler should be smart enough to handle it).

Reply:
And not add anything more to the docstring since we would not restrict the type at all.

Reply:
(Maybe could also check that the model is a supervised model.)

Reply:
I thought about this but lean against it because
1) AFAICT, that MLJ implements `StatisticalTraits.prediction_type` and how it does it is undocumented.
2) It's possible to get e.g. a `:probabilistic` and `:supervised` without being a `Probabilistic`. e.g. `ProbabilisticSupervisedDetector` does probabilistic outlier detection, but I suspect these models are not useful for us.
3) Why dispatch on values when we can dispatch on types?

Reply:
> Why dispatch on values when we can dispatch on types?

This point would actually be strongest argument in favour of traits: We would also support models that are not subtypes of `Probabilistic` or `Deterministic` (for whatever reason) and that we do not know about but whose predictions would still be of the desired form (probabilistic or deterministic).

Reply:
`prediction_type` shows up in the model search and if you print `info(model)`: https://alan-turing-institute.github.io/MLJ.jl/dev/model_search/ So it seems to be quite official?

Reply:
> This point would actually be strongest argument in favour of traits: We would also support models that are not subtypes of `Probabilistic` or `Deterministic` (for whatever reason) and that we do not know about but whose predictions would still be of the desired form (probabilistic or deterministic).

I think this presumes that the results of `fit` and `predict` for such models would be of the same form as we need. But it's not clear to me from the docs that they are (e.g. https://alan-turing-institute.github.io/MLJ.jl/dev/adding_models_for_general_use/#Outlier-detection-models). We would be relying on an undocumented trait implementation, where the decisions about which traits apply are not defined anywhere. In particular, for supervised outlier detection, it's not clear to me whether these models support multiple labels as we have. To be convinced, I'd need to see a test case of one of these detectors (see https://github.com/OutlierDetectionJL/OutlierDetection.jl) used to compute rstar.

Reply:
> So it seems to be quite official?

There's even an example that filters on `prediction_type` on the same page:
```julia
julia> filter(model) = model.is_supervised &&
                       model.input_scitype >: MLJ.Table(Continuous) &&
                       model.target_scitype >: AbstractVector{<:Multiclass{3}} &&
                       model.prediction_type == :deterministic
```

Reply:
It seems there are large portions of their traits interface that are undocumented.

I wonder if instead of using `prediction_type` it is more useful to use `input_scitype` to check that the models accept tables of continuous values, `target_scitype` to check that the model supports multiclass labels, and then `predict_scitype` to determine whether the predictions are labels, probabilities, or something else (error).

These have the benefit that the `scitype` interface is documented (mostly; not `predict_scitype`: https://alan-turing-institute.github.io/MLJ.jl/dev/adding_models_for_general_use/#Trait-declarations) and connects directly to what we need (we assume the model accepts certain inputs and labels and makes certain types of predictions).

Reply:
I think that's a good idea but unfortunately it seems that `Pipeline`s do not support the traits properly (maybe they can't - even though when working with instances all information should be available?):
```julia
julia> using MLJBase, MLJXGBoostInterface, MLJModelInterface

julia> const MMI = MLJModelInterface

julia> classifier = XGBoostClassifier();

julia> MMI.input_scitype(classifier)
Table{<:AbstractVector{<:Continuous}}

julia> MMI.target_scitype(classifier)
AbstractVector{<:Finite} (alias for AbstractArray{<:Finite, 1})

julia> MMI.predict_scitype(classifier)
AbstractVector{Density{<:Finite}} (alias for AbstractArray{ScientificTypesBase.Density{<:Finite}, 1})

julia> classifier = Pipeline(XGBoostClassifier(); operation=predict_mode);

julia> MMI.input_scitype(classifier)
Unknown

julia> MMI.target_scitype(classifier)
AbstractVector{<:Finite} (alias for AbstractArray{<:Finite, 1})

julia> MMI.predict_scitype(classifier)
Unknown
```

And supporting `Unknown` as well seems to be less satisfying since that is the fallback for models that don't implement the traits... I wonder if this could/should be fixed in MLJBase for `Pipeline`, and hence in principle could work for models that implement traits?

In any case, we could at least dispatch on the supported prediction types in `_rstar` instead of restricting it to specific model types.